libs/ui/src/components/NeoMessage/NeoMessage.vue

```diff
@@ -1,146 +1,137 @@
-        case 'is-info':
-          return 'information'
-        case 'is-success':
-          return 'check-circle'
-        case 'is-warning':
-          return 'alert'
-        case 'is-danger':
-          return 'alert-circle'
-        default:
-          return null
-      }
-    },
-  },
-  watch: {
-    active(value) {
-      this.isActive = value
-    },
-    isActive(value) {
-      if (value) {
-        this.setAutoClose()
-      } else {
-        if (this.timer) {
-          clearTimeout(this.timer)
-        }
-      }
-    },
-  },
-  mounted() {
-    this.setAutoClose()
-  },
-  methods: {
-    /**
-     * Close the Message and emit events.
-     */
-    close() {
-      this.isActive = false
-      this.resetDurationProgress()
-      this.$emit('close')
-      this.$emit('update:active', false)
-    },
-    click() {
-      this.$emit('click')
-    },
-    /**
-     * Set timer to auto close message
-     */
-    setAutoClose() {
-      if (this.autoClose) {
-        this.timer = setTimeout(() => {
-          if (this.isActive) {
-            this.close()
-          }
-        }, this.duration)
-      }
-    },
-    resetDurationProgress() {
-      /**
-       * Wait until the component get closed and then reset
-       **/
-      setTimeout(() => {
-        this.remainingTime = this.duration / 1000
-      }, 100)
-    },
+const iconVariant: Record<NeoMessageVariant, string> = {
+  info: 'circle-info',
+  success: 'check-circle',
+  warning: 'circle-exclamation',
+  danger: 'circle-exclamation',
+}
+
+const emit = defineEmits(['close', 'update:active', 'click'])
+const props = withDefaults(
+  defineProps<{
+    title?: string
+    active: boolean
+    closable: boolean
+    variant: NeoMessageVariant
+    autoClose: boolean
+    duration: number
+    showProgressBar: boolean
+  }>(),
+  {
+    active: true,
+    closable: true,
+    autoClose: false,
+    duration: 2000,
+    showProgressBar: false,
+    variant: 'success',
   },
+)
+
+const wrapper = ref()
+const isHovering = useElementHover(wrapper)
+
+const timer = ref()
+const isActive = ref(props.active)
+const remainingTime = ref(props.duration)
+
+const computedIcon = computed(() => iconVariant[props.variant] ?? null)
+
+const percent = computed(() => {
+  return (remainingTime.value / props.duration) * 100
+})
+
+const close = () => {
+  isActive.value = false
+  remainingTime.value = 0
+  emit('close')
+  emit('update:active', false)
 }
+
+watch(isActive, (active) => {
+  if (active) {
+    startTimer()
+  } else if (timer.value) {
+    clearTimeout(timer.value)
+  }
+})
+
+const timerTime = computed(() => (props.showProgressBar ? 100 : 1000))
+
+const startTimer = () => {
+  timer.value = setInterval(trackProgress, timerTime.value)
+}
+
+const trackProgress = () => {
+  if (remainingTime.value > 0 && !isHovering.value && props.showProgressBar) {
+    remainingTime.value -= timerTime.value
+  }
+}
+
+watch(remainingTime, (time) => {
+  if (time === 0 && props.autoClose) {
```

Comment:
Edge case: `duration` from props could be `1001`ms
```suggestion
  if (time <= 0 && props.autoClose) {
```

Reply:
thanks, fixed